atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweepThread.java

```diff
@@ -0,0 +1,300 @@
+        }
+    }
+
+    private Optional<TableToSweep> getTableToSweep() {
+            return specificTableSweeper.getTxManager().runTaskWithRetry(
+                    tx -> {
+                        Optional<SweepProgress> progress = specificTableSweeper.getSweepProgressStore()
+                                .loadProgress(threadIndex);
+                        SweepPriorityOverrideConfig overrideConfig = sweepPriorityOverrideConfig.get();
+                        if (progress.map(
+                                realProgress -> shouldContinueSweepingCurrentTable(realProgress, overrideConfig))
+                                .orElse(false)) {
+                            try {
+                                createLockForTableIfNecessary(progress.get().tableRef());
+                                lockForCurrentTable.lockOrRefresh();
+                                return Optional.of(TableToSweep.continueSweeping(progress.get().tableRef(),
+                                        lockForCurrentTable,
+                                        progress.get()));
+                            } catch (InterruptedException ex) {
+                                log.info("Sweep lost the lock for table {}",
+                                        LoggingArgs.tableRef(progress.get().tableRef()));
+                                // We'll fall through and choose a new table
+                            }
+                        }
+
+                        log.info("Sweep is choosing a new table to sweep.");
+                        return getNextTableToSweep(tx, overrideConfig);
+                    });
+    }
+
+    // This is needed if we've just grabbed the sweep thread lock and a previous thread was in the middle of a table
+    private void createLockForTableIfNecessary(TableReference tableRef) {
+        if (lockForCurrentTable != null) {
+            return;
+        }
+
+        lockForCurrentTable = SingleLockService.createNamedLockServiceForTable(
+                lockService, TABLE_LOCK_PREFIX, tableRef);
+    }
+
+    private boolean shouldContinueSweepingCurrentTable(
+            SweepProgress progress,
+            SweepPriorityOverrideConfig overrideConfig) {
+        String currentTableName = progress.tableRef().getQualifiedName();
+        if (overrideConfig.priorityTables().isEmpty()) {
+            return !overrideConfig.blacklistTables().contains(currentTableName);
+        }
+        return overrideConfig.priorityTables().contains(currentTableName);
+    }
+
+    private Optional<TableToSweep> getNextTableToSweep(Transaction tx, SweepPriorityOverrideConfig overrideConfig) {
+        Optional<TableToSweep> nextTableToSweep = nextTableToSweepProvider.getNextTableToSweep(
+                tx,
+                specificTableSweeper.getSweepRunner().getConservativeSweepTimestamp(),
+                overrideConfig);
+        nextTableToSweep.ifPresent(tableToSweep -> lockForCurrentTable = tableToSweep.getSweepLock());
```

Comment:
After the latest change, I think we actually should just keep the `TableToSweep`, rather than tracking a few separate things from it. We'd need to remember to update it each time with the latest progress, though (probably with a method on `TableToSweep`).

Will refactor after the per-table progress stuff has been checked.